For patch application specify that checksums need to be updated.

garretrieger opened the pull request:
Additionally, add an option for glyph keyed patches to use 24 bit glyph ids to be compatible with future font file extensions that allow for larger glyph ids.


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/w3c/IFT/pull/168.html" title="Last updated on Apr 26, 2024, 11:29 PM UTC (750b917)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/IFT/168/bbe8586...750b917.html" title="Last updated on Apr 26, 2024, 11:29 PM UTC (750b917)">Diff</a>

skef commented on Overview.bs:
```diff
@@ -1627,8 +1639,15 @@ The algorithm:
-6. Modify the mappings in the IFT and IFTX table(s) in <var>extended font subset</var> if present to remove any mapping entries
-    that map to this patch. <!-- TODO: this needs to be made more detailed once we have the mapping format sections -->
+6. Modify the mappings in the 'IFT ' or 'IFTX' table which has an [=Glyph keyed patch/id=] which matches <var>id</var> to remove any
+    entries that map to this patch. The entries to be removed can be determined by finding entries which have a URI that matches the
```
I'm not sure I understand this part. We're dealing with glyph-keyed patches in this section, correct? It sounds to me like we're saying the client should match the URI for the patch with the URI in the patch map, but isn't one of those "filled in" and one of them still a template? And then I guess the idea would be to derive the ID for the patch by extracting it during the matching process? If that's what's intended, surely no implementation would do that in practice. It would either stash the URI template and the ID used to fill in the template somewhere while waiting for the file to download, or (more likely) just stash the identity of the patch map and the ID (skipping the process of matching the template).

It sounds like we might be adapting the procedures for the other patch types to our glyph-keyed patches a little too directly. 

Or maybe I don't understand what's going on in this section of the documentation.

garretrieger replied:
So at a high level what I'm trying to communicate here is that the client will need to go set the bit in appliedBitmaps (or the equivalent 'ignored' bit in format 2) for the patch which is being applied. I'll take a pass at rewriting this to be more clear.

One important thing to note is that client implementations can choose to do things differently so long as the end result is the same (see: https://w3c.github.io/IFT/Overview.html#w3c-conformant-algorithms).  So for example with format 1 they could use entry index to quickly locate the bit to set, and for format2 they would probably want to record the position of the flags byte w/ the entry to quickly get at it later. For the spec that means we don't have to specify the most efficient way to perform an algorithm, and should instead bias towards the most clear explanation of what should happen. So in this case matching the patch's URL to the URL in the patch map entry explains which entries to update without having to dive into the details of the two different formats.

garretrieger commented on Overview.bs:
```diff
@@ -1446,15 +1444,23 @@ The algorithm outputs:
 
-3. For each entry in [=Per table brotli patch/patches=], with index <var>i</var>:
+
+4. In the following steps when a table is added to <var>extended font subset</var> then, table data is added to the font and a
```
Thanks, your rewrite is more clear. Just replaced it.

skef commented:
> For the spec that means we don't have to specify the most efficient way to perform an algorithm, and should instead bias towards the most clear explanation of what should happen. So in this case matching the patch's URL to the URL in the patch map entry explains which entries to update without having to dive into the details of the two different formats.

I don't disagree with this broadly speaking, but if I'm tracking this part of the system right we have (or *can* have) a URI template in one case and a filled-in URI in another case. In such a case I'm worried that the act of "matching" these two strings is at least vague and maybe problematic, and therefore doesn't provide the desired clarity.

garretrieger commented:
> > For the spec that means we don't have to specify the most efficient way to perform an algorithm, and should instead bias towards the most clear explanation of what should happen. So in this case matching the patch's URL to the URL in the patch map entry explains which entries to update without having to dive into the details of the two different formats.
> 
> I don't disagree with this broadly speaking, but if I'm tracking this part of the system right we have (or _can_ have) a URI template in one case and a filled-in URI in another case. In such a case I'm worried that the act of "matching" these two strings is at least vague and maybe problematic, and therefore doesn't provide the desired clarity.

At patch application time we always have a fully expanded patch URI which is produced during the "extending a font subset" algorithm. I just updated the application algorithm here to take that as an explicit input argument (and updated the caller to pass it). On the patch map side each entry can be assigned a fully expanded URI as well via the "Interpretting format X" algorithm. I don't expect any implementation to actually do that of course, but writing it this way succinctly identifies which entries need to be removed.

I took a pass at rewriting that step, is the new text more understandable to you?

skef commented:
> is the new text more understandable to you?

The honest immediate answer is "not enough to make a difference", but this might just be because I need to re-read some of the earlier sections more carefully. 

Maybe I don't understand the layers of data in the map. The text indicates I've already identified the right table for the ID, why do I then need to look at URIs? Can a single table (IFT or IFTX) now have both glyph-based patches and other patches simultaneously? Or two types of glyph-based patches? If not, doesn't the ID we've *just* talked about pick out the bit that needs to be changed pretty directly? 

> On the patch map side each entry can be assigned a fully expanded URI as well via the "Interpretting format X" algorithm. I don't expect any implementation to actually do that of course, but writing it this way succinctly identifies which entries need to be removed.

OK, but I just read through that section and I don't find the association between the theoretical expanded table you describe and the actual bit that needs to be changed to be very strong. You seem to be saying to examine the expanded table you describe, find the right entry, and then modify the bit not in that expanded table, but in the original table that was derived from it. But I don't see any actual association between the expanded entry and the bit, you just seem to be relying on an understand of how the expanded table is derived. This is abstract to the point of difficult to follow. If you made that association more explicit in the original description, maybe adding a "field" to describe the bit *in those terms* I would find all of this easier to follow. 

(I've commented a couple times before that when I encounter the "interpreting" section in the spec at the point where it occurs I start out having no idea what it's talking about or telling me to do, and then I have to try to remember the conversations we've had about it and piece it together. Maybe it achieves a kind of succinctness but I still think we have a problem in the clarity department.)

skef commented:
Maybe my problem boils down to this:

> <h5 algorithm id="interpreting-patch-map-format-2">Interpreting Format 2</h5>

> This algorithm is used to convert a format 2 patch map into a list of [=patch map entries=].

As I read it there is one word in this section indicating that this is a way of understanding the contents of a patch table rather than "converting" it to a different format, which is "interpreting" in the heading. In my view headings are ideally "extra" -- they summarize the text in the section. If as a reader I ignore the heading, I have a procedure for translating one format to an expanded format. 

We seem to be asking the reader to 1) do this translation mentally and 2) let the translated entry stand for the original entry, such that they're sort of equivalent. Thus we talk of modifying a bit in an entry that, if we were taking this literally, doesn't exist, because the bit was in the original and not the new one. And that whole way of looking at things rests on the word "interpreting" in the heading. 

I'm still not sure this is the right way to go, but if it is I really think we need a paragraph actually describing what you want the reader to do mentally in relation to it.

skef commented:
I guess, just to pull out a specific issue here, the "Interpreting Format 1" section mentions the `appliedEntriesBitMap` field twice, both to indicate that an entry should be skipped if the corresponding bit is 1. I understand abstractly how "setting the bit" results in "removing the entry", but what is the metaphor here? If I'm "converting" the map, wouldn't I need to "re-convert" it for the entries to be "removed"? So this isn't really a "conversion" after all: the virtual map changes immediately as the source map changes.

I worry this will sound like nitpicking, but I'm not trying to model what someone else *might* think on encountering this text, I'm describing the process *I'm* going through in trying to understand what the text is actually talking about. And having done that I can see what you're getting at.

But not only do I feel is this too hard, it seems like in this case the whole priority has been inverted. As I asked earlier, if we can already openly talk about the ID, why are we consulting the virtual translated map at all? It's not to figure out what the ID is by matching the URL. It's not to figure what the bit we need to set is -- that bit isn't "there". So it seems like we're talking about the translated format in order to note how the entries are "removed" in virtue of setting the bits. That is, we appear to be doing bookkeeping on a virtual map in order to clarify its virtual properties. So *locally*, at least, this is not *succinct*. That doesn't mean it isn't *globally* succinct, but if the reader needs this sort of support to understand what's going on, it at least calls that into question.

garretrieger commented:
> I guess, just to pull out a specific issue here, the "Interpreting Format 1" section mentions the `appliedEntriesBitMap` field twice, both to indicate that an entry should be skipped if the corresponding bit is 1. I understand abstractly how "setting the bit" results in "removing the entry", but what is the metaphor here? If I'm "converting" the map, wouldn't I need to "re-convert" it for the entries to be "removed"? So this isn't really a "conversion" after all: the virtual map changes immediately as the source map changes.
> 

In the context of the "extending a font subset" algorithm reconversion after each modification is how it works, roughly:

1. font_subset = the inital font subset.
2. Decode the patch map(s) of font_subset (via the interpret algorithm).
3. Pick out a patch from the decoded map, load, and apply it. This produces a new font subset, set font_subset to it. Go back to 2. (this repeats until there are no patches left to be applied)

So on each iteration of the algorithm the encoded patch map is modified and then re-interpretted via the interpret algorithm to realize the changes to the virtual representation. An efficient client implementation would side step 2 and when removing entries remove them from both the font subset bytes and from it's current virtual map.

> I worry this will sound like nitpicking, but I'm not trying to model what someone else _might_ think on encountering this text, I'm describing the process _I'm_ going through in trying to understand what the text is actually talking about. And having done that I can see what you're getting at.
> 
> But not only do I feel is this too hard, it seems like in this case the whole priority has been inverted. As I asked earlier, if we can already openly talk about the ID, why are we consulting the virtual translated map at all? It's not to figure out what the ID is by matching the URL. It's not to figure what the bit we need to set is -- that bit isn't "there". So it seems like we're talking about the translated format in order to note how the entries are "removed" in virtue of setting the bits. That is, we appear to be doing bookkeeping on a virtual map in order to clarify its virtual properties. So _locally_, at least, this is not _succinct_. That doesn't mean it isn't _globally_ succinct, but if the reader needs this sort of support to understand what's going on, it at least calls that into question.

I think there might be some confusion about the "ID" mentioned in this step since we have multiple different IDs in play. Here the ID that's mentioned is the unique ID of the table. All glyph keyed patches listed in that table will have the same ID inside of them, since the ID in the patch must match the table ID to be applied. So we can't use the ID to identify a specific patch. Naming it just "ID" is definitely kind of ambiguous, so I'll rename to be more clear what it's for. For uniquely identifying patches there's two options:
1. The numeric entry index: both formats have the concept of a numeric entry index. In format 1 this is unique, however in format 2 it might not be because of the application of the index to string map (consider a case where two different indices map to the same string). This situation is likely rare in practice but isn't invalid. The extension algorithm will still work correctly, but once a patch from a specific url has been applied then all entries that reference that same URL should be removed since we don't want the same patch to be applied twice.
2. So that leaves us with the second option the patch URI which is unique.

So we'll need to use the URI, but based on your feedback there's definitely some additional changes that would probably be good:
1. I agree that that connection from the virtual map entries back to the actual physical bit that needs to be set is  ambigous/underspecified at the moment, so I'll add a short "remove an entry" algorithm to each of the patch map formats that will provide a more complete description of how go from URI to the actual bit that needs to be set.
2. In the introduction to patch maps I'll add explanation on how the interpreting algorithms and the output virtual maps ties into the whole process.

garretrieger commented:
Another change we could consider, is reordering of the sections. Right now I've set things up so we introduce all of the individual pieces and then tie them together in the end with the "Extending a font subset" section, but maybe it might aid understanding if we actually start with the extension algorithm which may help give the big picture before then introducing the specifics of the patch mapping and patch formats.

svgeesus commented:
@garretrieger could you rebase your branch to get rid of the Overview.html conflicts?

garretrieger commented:
Ok, I've made some edits:
1. Added algorithms defining how to remove entries for the two patch map formats. The glyph keyed patch application step now invokes these.
2. Updated the patch map introduction with a paragraph that explains how the patch map data flows through from the encoded bytes to extension algorithm and patch applications.
3. Renamed table "ID" to "compatibility ID" to disambiguate it from other IDs in the spec.

Can you take another look?

skef commented on Overview.bs:
```diff
@@ -1036,6 +1071,27 @@ The algorithm:
+
+* <var>patch URI</var>: URI for a patch which identifies the entries to be removed.
+
+This algorithm is a modified version of [$Interpret Format 2 Patch Map$], invoke [$Interpret Format 2 Patch Map$] with <var>patch map</var>
```
The patch 1 removal description reads quite clear to me now. This still has some of the conceptual/editorial aspects that concern me. We don't have to deal with all of that in this PR but I may still want to revisit it.

garretrieger replied:
Sounds good

garretrieger commented on Overview.bs:
```diff
@@ -445,7 +451,7 @@ of the patch map table. A map is encoded in one of two formats:
   <tr>
     <td>uint32</td>
-    <td><dfn for="Format 1 Patch Map">id</dfn>[4]</td>
+    <td><dfn for="Format 1 Patch Map">compatibilityId</dfn>[4]</td>
```
It's hard to read with the markup, but it's still 128 bits there's a [4] after indicating it's an array of 4 uint32's. Agreed, we don't actually talk about how to set this anywhere and should definitely provide some guidance.

garretrieger replied:
Should be a smallish change so I'll add that to this PR.

skef replied:
I should add: Maybe even more important than preventing the attempted integration of a patch from a different font, this field was originally intended to prevent attempted integration of a patch from a *different encoding* of the *same* font. I've run into that in practice a number of times even with my very limited use of the system, and having the ID match fail is far preferable to spending an hour wondering if the system has a bug before realizing that the files don't match. It can happen when you overwrite one file but not another, and it can also happen because of caches that don't get cleared.

skef replied:
Oh, my bad. (I wrote that second comment before I saw yours -- github glitch).

skef replied:
Sorry, my head has been fuzzy for the past few days. Normally I'm better at not going on these mistaken detours.

skef replied:
I think I picked 128 because it was the number where, assuming you have a decent random number generator with a real seed, you don't have to worry about manually checking for collisions anymore. It might be worth putting something to that effect in a "Note".

garretrieger replied:
Done.

garretrieger replied:
Added a bit of extra text to avoid collisions within a particular IFT font.
